vertical-pod-autoscaler/pkg/admission-controller/logic/limit_range_calculator.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logic
+
+import (
+	"fmt"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+
+	v1_listers "k8s.io/client-go/listers/core/v1"
+)
+
+// LimitsRangeCalculator checks for LimitRange and if container needs limits to be set
+type LimitsRangeCalculator interface {
+	// GetContainerLimitRangeItem returns LimitRangeItem that describes limitation on container limits in the given namespace.
+	GetContainerLimitRangeItem(namespace string) (*v1.LimitRangeItem, error)
+}
+
+type noopLimitsRangeCalculator struct{}
+
+func (lc *noopLimitsRangeCalculator) GetContainerLimitRangeItem(namespace string) (*v1.LimitRangeItem, error) {
+	return nil, nil
+}
+
+type limitsChecker struct {
+	limitRangeLister v1_listers.LimitRangeLister
+}
+
+// NewLimitsRangeCalculator returns a limitsChecker or an error it encountered when attempting to create it.
+func NewLimitsRangeCalculator(f informers.SharedInformerFactory) (*limitsChecker, error) {
+	if f == nil {
+		return nil, fmt.Errorf("NewLimitsRangeCalculator requires a SharedInformerFactory but got nil")
+	}
+	limitRangeLister := f.Core().V1().LimitRanges().Lister()
+	stopCh := make(chan struct{})
+	f.Start(stopCh)
+	for _, ok := range f.WaitForCacheSync(stopCh) {
+		if !ok {
+			if f.Core().V1().LimitRanges().Informer().HasSynced() {
+				return nil, fmt.Errorf("informer did not sync")
+			}
+		}
+	}
+	return &limitsChecker{limitRangeLister}, nil
+}
+
+// NewNoopLimitsCalculator returns a limit calculator that instantly returns no limits.
+func NewNoopLimitsCalculator() *noopLimitsRangeCalculator {
+	return &noopLimitsRangeCalculator{}
+}
+
+func (lc *limitsChecker) GetContainerLimitRangeItem(namespace string) (*v1.LimitRangeItem, error) {
+	limitRanges, err := lc.limitRangeLister.LimitRanges(namespace).List(labels.Everything())
+	if err != nil {
+		fmt.Printf("1")
+		return nil, fmt.Errorf("error loading limit ranges: %s", err)
+	}
+
+	for _, lr := range limitRanges {
+		for _, lri := range lr.Spec.Limits {
+			if lri.Type == v1.LimitTypeContainer || lri.Type == v1.LimitTypePod && (lri.MaxLimitRequestRatio == nil || lri.Default == nil) {
```

Comment:
Hi @jbartosik , I don't get this. In case a LimitRange has MaxRatio and Default both not nil we don't choose that Limit?
Also, in server code I understand that [any limit range have to be respected](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/limitranger/admission.go#L106) so the more restrictive (minumum max ratio) have to be respected. In this code I understand that we are not checking for the more restrictive limits max ratio.

Reply:
My mistake. We care about limit range if it has at least one of Max / Default set.

Reply:
but what about the MaxRatio? the purpose of my initial PR was exactly just to shrink limits in case a max ratio was present in LimitRange definition to prevent that after the vpa-updater kill a pod the pod couldn't be recreated anymore because the vpa-admission-controller is setting the requests too low and the LimitRanger admission controller prevent the pod to be created.
I am sorry but I am a bit lost on this rework. this PR like it is now doesn't address anymore the issue https://github.com/kubernetes/autoscaler/issues/1812

Reply:
@safanaj since your PR we have changed the logic of VPA to keep the ratio between request and limit constant. This means that VPA will never break things - since we keep the original request to limit ratio that was specified for the deployment, we rely on the user to specify it in a way that respects the MaxRatio.